README.md
# Comidas - Ejercicio de modelado en Haskell

[![Build Status](https://travis-ci.com/uqbar-project/comidas-haskell.svg?branch=master)](https://travis-ci.com/uqbar-project/comidas-haskell)

## Enunciado

Queremos que la persona pueda comer distintas comidas: existen las ensaladas, las hamburguesas y las paltas, quiero comer las distintas cosas y cada cosa aporta distinto al comerla.

- una ensalada de x kilos aporta la mitad de peso para la persona y no agrega colesterol
- cada hamburguesa tiene una cantidad de ingredientes: el colesterol aumenta un 50% para la persona y lo hace engordar en (3  * la cantidad de ingredientes) kilos
- la palta aumenta 2 kilos a quien la consume (el colesterol que aporta es despreciable)

## Alternativa 1: la comida es una función

```hs
type Comida = Persona -> Persona

ensalada :: Kilos -> Comida
ensalada kilos persona = persona {
  peso = peso persona + (kilos / 2)
}

hamburguesa :: [Ingrediente] -> Comida
hamburguesa ingredientes persona = persona {
  peso = peso persona + (3 * length ingredientes),
  colesterol = colesterol persona * 1.5
}

palta :: Comida
palta persona = persona {
  peso = peso persona + 2
}
```

#### ¡Quiero mi menú!
¿Puedo tener una lista de comidas? Sí, claro:

```hs
[ensalada 1, hamburguesa ["cheddar", "bacon"], palta, ensalada 3][<una función>,<una función>,<una función>,<una función>]
```

Y puedo pedirle a una persona que coma un conjunto de comidas:

```hs
almuerzo :: [Comida]
almuerzo = [ensalada 1, hamburguesa ["cheddar", "bacon"], palta, ensalada 3]

almorzar :: Persona -> Persona
almorzar persona = foldr ($) persona almuerzo
```

#### Pero pero peeeeero
Comer es una acción intrínseca a la comida, no puedo hacer otra cosa: las comidas no se pueden comparar, ni podemos acceder a información de una comida (ingredientes, o calorías que genera).

#### ¿Ventajas?
La ventaja principal es en lo didáctico:
- Modelar con funciones ¡es pensar en funcional!
- Además, para poder comer necesitamos tener aplicación parcial y orden superior, así que de paso estamos repasando esos temas.

#### ¿Y el locro?
Cuando aparece una nueva comida, eso no afecta lo que construimos anteriormente, solo se escribe una nueva función y eso nos sirve para incorporarla a cualquier almuerzo.

## Segunda variante: múltiples constructores

Si pensamos que la comida se modela como un tipo de dato con múltiples constructores, donde cada comida necesita un constructor diferente, tenemos esta solución:

```hs
data Comida' = Ensalada Kilos | Hamburguesa [Ingrediente] | Palta
  deriving (Eq, Ord, Show)

comer :: Comida' -> Persona -> Persona
comer (Ensalada kilos) persona = persona {
  peso = peso persona + (kilos / 2)
}

comer (Hamburguesa ingredientes) persona = persona {
  peso = peso persona + (3 * length ingredientes),
  colesterol = colesterol persona * 1.5
}

comer (Palta) persona = persona {
  peso = peso persona + 2
}
```

#### ¿Ventajas?
Separamos aquí el dato comida vs. la acción de comer. Tenemos información sobre la comida (kilos, ingredientes), podemos comparar las comidas, e incluso tener acciones diferentes (comer, endulzar, mejorar, etc.)

#### ¡Quiero mi menú!
Podemos tener una lista de comidas también: 

```hs
[Ensalada 1, Hamburguesa ["cheddar", "bacon"], Palta, Ensalada 3]
```

Y para almorzar, necesito aplicar la función comer al dato de la comida:

```hs
almorzar' :: Persona -> Persona
almorzar' persona = foldr comer persona almuerzo'
```

#### Pero pero peeeeero
No aparece la aplicación parcial (estás construyendo valores simplemente), el orden superior necesita que no construyas una función recursiva donde solamente apliques la función comer.

#### ¿Y el locro?
Cuando aparece una nueva comida, eso requiere ir a modificar cada una de las funciones que trabajan sobre la comida: comer, endulzar, mejorar, etc. Esto no es una desventaja, sino algo que se desglosa de la ventaja de poder tener distintas operaciones con las comidas y no sólo la operación "comer", como en el caso de modelar con funciones. 

Más información en [este artículo](http://wiki.uqbar.org/wiki/articles/data--definiendo-nuestros-tipos-en-haskell.html)

## Alternativa 3: typeclasses

La variante más tirada de los pelos consiste en pensar en la comida como un typeclass:

```hs
class Comida'' a where
   comer'' :: a -> Persona -> Persona

data Ensalada'' = Ensalada'' Kilos
  deriving (Eq, Ord, Show)

data Hamburguesa'' = Hamburguesa'' [Ingrediente]
  deriving (Eq, Ord, Show)

data Palta'' = Palta''
  deriving (Eq, Ord, Show)

instance Comida'' Ensalada'' where
  comer'' (Ensalada'' kilos) persona = persona {
    peso = peso persona + (kilos / 2)
  }

instance Comida'' Hamburguesa'' where
  comer'' (Hamburguesa'' ingredientes) persona = persona {
    peso = peso persona + (3 * length ingredientes),
    colesterol = colesterol persona * 1.5
  }

instance Comida'' Palta'' where
  comer'' (Palta'') persona = persona {
    peso = peso persona + 2
  }
```

#### ¿Y el locro?
Agregar una nueva comida se hace fácilmente: se agrega un `data Locro` y después un `instance Comida Locro`, implementando el `comer''` como en el ejemplo de arriba. Incluso se puede agregar el Locro en otro archivo, como al modelar con funciones.

#### PEERO PEERO PEEEEEERO ¡No es un tipo!
Ojo, porque la comida **no es un tipo**, sino un agrupador de diferentes tipos. Esto significa que la palta podría tener funciones específicas mientras que la ensalada podría tener otras. Si queremos tener funciones para cada tipo de comida, debemos escribirlas en el typeclass. Una ventaja sería que podríamos definir una función en la typeclass en base a otra:

```hs
class Comida'' a where
   comer'' :: a -> Persona -> Persona

   repetir'' :: a -> Persona -> Persona
   repetir'' comida persona = (comer'' comida . comer'' comida) persona
```

Cada comida necesita un constructor específico, que es propio para el tipo: el constructor Ensalada define el tipo Ensalada, y nada más. 

#### Pero pero peeeero... No tengo menú
El primer problema (la comida no es un tipo) se vuelve evidente cuando queremos tener una lista de comidas:

```hs
comidas = [Palta'', Ensalada'' 5, Hamburguesa'' ["cheddar"]]

<interactive>:14:21: error:
    • Couldn't match expected type ‘Palta''’
                  with actual type ‘Ensalada''’
    • In the expression: Ensalada'' 5
      In the expression:
        [Palta'', Ensalada'' 5, Hamburguesa'' ["cheddar"]]
      In an equation for ‘comidas’:
          comidas = [Palta'', Ensalada'' 5, Hamburguesa'' ["cheddar"]]

<interactive>:14:35: error:
    • Couldn't match expected type ‘Palta''’
                  with actual type ‘Hamburguesa''’
    • In the expression: Hamburguesa'' ["cheddar"]
      In the expression:
        [Palta'', Ensalada'' 5, Hamburguesa'' ["cheddar"]]
      In an equation for ‘comidas’:
          comidas = [Palta'', Ensalada'' 5, Hamburguesa'' ["cheddar"]]
```

es decir, colisionan la palta, la ensalada y la hamburguesa porque son de diferente tipo. Tampoco puedo compararlas, el operador (==) necesita que los elementos a comparar sean del mismo tipo:

```hs
(==) :: Eq a => a -> a -> a
```

Por más que la palta y la ensalada deriven de Eq, no podemos asociarlas a variables del mismo tipo. Y lo propio pasa con los operadores de Ord: (<), (>), etc.

El segundo problema es que estamos queriendo utilizar un concepto que fue pensado para otra cosa, que [no tiene que ver con modelar datos](https://www.reddit.com/r/haskell/comments/1j0awq/definitive_guide_on_when_to_use_typeclasses/). Incluso hay gente que piensa que es un antipattern hacer esto: [Haskell Antipattern: Existential Typeclass](https://lukepalmer.wordpress.com/2010/01/24/haskell-antipattern-existential-typeclass/)

En definitiva, los typeclasses permiten agrupar tipos diferentes, y a nosotros nos conviene pensar en compartir todas las comidas con el mismo tipo.

## Resumen

Hemos comparado las tres soluciones posibles para modelar comidas: si bien utilizar múltiples constructores es la opción que hoy suele usarse comercialmente en la programación funcional, hay motivos didácticos muy fuertes para modelar con funciones aquellas abstracciones sabiendo que tenemos la restricción de definir una única operación principal y que no nos interesa conocer información relacionada, a favor de trabajar más fuertemente los conceptos de orden superior y aplicación parcial.


ChangeLog.md
# Changelog for comidas-haskell

## Unreleased changes


test/Spec.hs
import PdePreludat
import Test.Hspec

main :: IO ()
main = hspec $ do
  describe "Test de ejemplo" $ do
    it "El pdepreludat se instaló correctamente" $ do
      (1 + 1) `shouldBe` 2



src/LibraryData.hs
module LibraryData where
import PdePreludat

type Kilos = Number
type Ingrediente = String

data Persona = Persona {
  colesterol :: Number,
  peso :: Kilos
}

-- Queremos que la persona pueda comer distintas comidas. 
-- Existen las ensaladas, las hamburguesas y las paltas, 
-- quiero comer las distintas cosas y cada cosa aporta distinto 
-- al comerla

-- una ensalada de x kilos aporta la mitad de peso para la persona
--     y no agrega colesterol
data Comida = Ensalada Kilos | Hamburguesa [Ingrediente] | Palta
  deriving (Eq, Ord, Show)

comer :: Comida -> Persona -> Persona
comer (Ensalada kilos) persona = persona {
  peso = peso persona + (kilos / 2)
}

-- cada hamburguesa tiene una cantidad de ingredientes
-- el colesterol aumenta un 50%
-- el peso aumenta en 3 kilos * la cantidad de ingredientes
comer (Hamburguesa ingredientes) persona = persona {
  peso = peso persona + (3 * length ingredientes),
  colesterol = colesterol persona * 1.5
}

-- la palta aumenta 2 kilos a quien la consume
comer (Palta) persona = persona {
  peso = peso persona + 2
}



src/LibraryFunction.hs
module LibraryFunction where
import PdePreludat

type Kilos = Number
type Ingrediente = String

data Persona = Persona {
  colesterol :: Number,
  peso :: Kilos
}

-- Queremos que la persona pueda comer distintas comidas. 
-- Existen las ensaladas, las hamburguesas y las paltas, 
-- quiero comer las distintas cosas y cada cosa aporta distinto 
-- al comerla

-- una ensalada de x kilos aporta la mitad de peso para la persona
--     y no agrega colesterol
type Comida = Persona -> Persona

ensalada :: Kilos -> Comida
ensalada kilos persona = persona {
  peso = peso persona + (kilos / 2)
}

-- cada hamburguesa tiene una cantidad de ingredientes
-- el colesterol aumenta un 50%
-- el peso aumenta en 3 kilos * la cantidad de ingredientes
hamburguesa :: [Ingrediente] -> Comida
hamburguesa ingredientes persona = persona {
  peso = peso persona + (3 * length ingredientes),
  colesterol = colesterol persona * 1.5
}

-- la palta aumenta 2 kilos a quien la consume
palta :: Comida
palta persona = persona {
  peso = peso persona + 2
}



src/LibraryTypeclasses.hs
module LibraryTypeclasses where
import PdePreludat

type Kilos = Number
type Ingrediente = String

data Persona = Persona {
  colesterol :: Number,
  peso :: Kilos
}

-- Queremos que la persona pueda comer distintas comidas. 
-- Existen las ensaladas, las hamburguesas y las paltas, 
-- quiero comer las distintas cosas y cada cosa aporta distinto 
-- al comerla

-- una ensalada de x kilos aporta la mitad de peso para la persona
--     y no agrega colesterol
class Comida a where
   comer :: a -> Persona -> Persona

-- una ensalada de x kilos aporta la mitad de peso para la persona
--     y no agrega colesterol
data Ensalada = Ensalada Kilos
  deriving (Eq, Ord, Show)

data Hamburguesa = Hamburguesa [Ingrediente]
  deriving (Eq, Ord, Show)

data Palta = Palta
  deriving (Eq, Ord, Show)

instance Comida Ensalada where
  comer (Ensalada kilos) persona = persona {
    peso = peso persona + (kilos / 2)
  }

-- cada hamburguesa tiene una cantidad de ingredientes
-- el colesterol aumenta un 50%
-- el peso aumenta en 3 kilos * la cantidad de ingredientes
instance Comida Hamburguesa where
  comer (Hamburguesa ingredientes) persona = persona {
    peso = peso persona + (3 * length ingredientes),
    colesterol = colesterol persona * 1.5
  }

-- la palta aumenta 2 kilos a quien la consume
instance Comida Palta where
  comer (Palta) persona = persona {
    peso = peso persona + 2
  }

-- comidas = [Palta, Ensalada 5, Hamburguesa 1]

-- • Couldn't match expected type ‘Palta’ with actual type ‘Ensalada’
-- • In the expression: Ensalada 5
--   In the expression: [Palta, Ensalada 5, Hamburguesa 1]
--   In an equation for ‘comidas’:
--       comidas = [Palta, Ensalada 5, Hamburguesa 1]

src/Library.hs
module Library
where

--import PdePreludat
import Data.List
type Number = Float
--

type Kilos = Number

type Ingrediente = String

data Persona = Persona {
  colesterol :: Number,
  peso :: Kilos
} deriving Show

sobrepeso::Persona -> Bool
sobrepeso alguien= peso alguien > 100

grandote,chiquito::Persona
grandote = Persona 150 150
chiquito = Persona 60 60

--1) Queremos que la persona pueda comer distintas comidas. 
-- Existen las ensaladas, las hamburguesas y las paltas, 
-- quiero comer las distintas cosas y cada cosa aporta distinto 
-- al comerla

-- una ensalada de x kilos aporta la mitad de peso para la persona
--     y no agrega colesterol
type Comida = Persona -> Persona

ensalada :: Kilos -> Comida
ensalada kilos persona = persona {
  peso = peso persona + (kilos / 2)
}

-- cada hamburguesa tiene una cantidad de ingredientes
-- el colesterol aumenta un 50%
-- el peso aumenta en 3 kilos * la cantidad de ingredientes
hamburguesa :: [Ingrediente] -> Comida
hamburguesa ingredientes persona = persona {
  peso = peso persona + (3 * genericLength ingredientes),
  colesterol = colesterol persona * 1.5
}

-- la palta aumenta 2 kilos a quien la consume
palta :: Comida
palta persona = persona {
  peso = peso persona + 2
}

almuerzo :: [Comida]
almuerzo = [ensalada 1, hamburguesa ["cheddar", "bacon"], palta, ensalada 3]

almorzar :: Persona -> Persona
almorzar persona = foldr ($) persona almuerzo

-- para mayor expresividad
-- almorzar persona = foldr comer persona almuerzo
-- comer persona comida = comida persona

--2) Queremos que todas las comidas se puedan comer dos veces seguidas
repetir:: Comida -> Persona -> Persona
repetir comida = comida.comida 

--3) Queremos ver si un almuerzo contiene una comida dada
--contieneComida::Comida -> [Comida]->Bool
-- No podemos hacerlo porque las comidas no son Eq

--4) Queremos averiguar si una comida es sabrosa.
-- las ensaladas son sabrosas cuando tienen más de un kilo
-- las hamburguesas son sabrosas cuando tienen cheddar
-- las paltas son sabrosas
-- No podemos hacerlo

--5) Queremos averiguar si una comida va a ser disfrutada por alguien.
-- Para quienes pesan más de 100 kg, todas las comidas son disfrutadas, 
-- para los demás, solo son disfrutadas las comidas sabrosas
-- Necesitamos nuevas funciones para disfrutar cada comida 
-- (tal vez convenga renombrar las anteriores, x ej comerEnsalada en vez de ensalada)
disfrutarEnsalada :: Kilos -> Persona -> Bool
disfrutarEnsalada kilos alguien = sobrepeso alguien || kilos > 1

disfrutarHamburguesa :: [Ingrediente] -> Persona -> Bool
disfrutarHamburguesa ingredientes alguien = sobrepeso alguien || elem "cheddar" ingredientes

disfrutarPalta:: Persona -> Bool
disfrutarPalta _ = True

-- ************************************************************
-- CON DATA
-- ************************************************************

-- Queremos que la persona pueda comer distintas comidas. 
-- Existen las ensaladas, las hamburguesas y las paltas, 
-- quiero comer las distintas cosas y cada cosa aporta distinto 
-- al comerla

-- una ensalada de x kilos aporta la mitad de peso para la persona
--     y no agrega colesterol
data Comida' = Ensalada Kilos | Hamburguesa [Ingrediente] | Palta
  deriving (Eq, Ord, Show)

comer' :: Comida' -> Persona -> Persona
comer' (Ensalada kilos) persona = persona {
  peso = peso persona + (kilos / 2)
}

-- cada hamburguesa tiene una cantidad de ingredientes
-- el colesterol aumenta un 50%
-- el peso aumenta en 3 kilos * la cantidad de ingredientes
comer' (Hamburguesa ingredientes) persona = persona {
  peso = peso persona + (3 * genericLength ingredientes),
  colesterol = colesterol persona * 1.5
}

-- la palta aumenta 2 kilos a quien la consume
comer' Palta persona = persona {
  peso = peso persona + 2
}


almuerzo' :: [Comida']
almuerzo' = [Ensalada 1, Hamburguesa ["cheddar", "bacon"], Palta, Ensalada 3]

almorzar' :: Persona -> Persona
almorzar' persona = foldr comer' persona almuerzo'

--2) Queremos que todas las comidas se puedan comer dos veces seguidas
repetir':: Comida' -> Persona -> Persona
repetir' comida persona = (comer' comida.comer' comida) persona

--3) Queremos ver si un almuerzo contiene una comida dada
contieneComida'::Comida'->[Comida']->Bool
contieneComida' comida comidas = elem comida comidas
--contieneComida = elem

--4) Queremos averiguar si una comida es sabrosa.
sabrosa' :: Comida' -> Bool

-- las ensaladas son sabrosas cuando tienen más de un kilo
sabrosa' (Ensalada kilos) = kilos > 1

-- las hamburguesas son sabrosas cuando tienen cheddar
sabrosa' (Hamburguesa ingredientes) = elem "cheddar" ingredientes

--las paltas son sabrosas
sabrosa' Palta = True

--5) Queremos averiguar si una comida va a ser disfrutada por alguien.
-- Para quienes pesan más de 100 kg, todas las comidas son disfrutadas, 
-- para los demás, solo son disfrutadas las comidas sabrosas
disfrutar' :: Comida' -> Persona -> Bool
disfrutar' comida alguien = sobrepeso alguien || sabrosa' comida




-- ************************************************************
-- CON TYPECLASSES
-- ************************************************************
--1) Queremos que la persona pueda comer distintas comidas. 
-- Existen las ensaladas, las hamburguesas y las paltas, 
-- quiero comer las distintas cosas y cada cosa aporta distinto 
-- al comerla

-- una ensalada de x kilos aporta la mitad de peso para la persona
--     y no agrega colesterol
class Comida'' a where
   comer'' :: a -> Persona -> Persona

   repetir'' :: a -> Persona -> Persona
   repetir'' comida persona = (comer'' comida . comer'' comida) persona

   sabrosa'' :: a -> Bool

data Ensalada'' = Ensalada'' {kilos::Kilos}
  deriving (Eq, Ord, Show)

data Hamburguesa'' = Hamburguesa'' {ingredientes::[Ingrediente]}
  deriving (Eq, Ord, Show)

data Palta'' = Palta''
  deriving (Eq, Ord, Show)

-- una ensalada de x kilos aporta la mitad de peso para la persona
--     y no agrega colesterol
instance Comida'' Ensalada'' where
  comer'' ensalada persona = persona {
    peso = peso persona + (kilos ensalada / 2)
  }
-- las ensaladas son sabrosas cuando tienen más de un kilo
  sabrosa'' ensalada = kilos ensalada > 1

-- cada hamburguesa tiene una cantidad de ingredientes
-- el colesterol aumenta un 50%
-- el peso aumenta en 3 kilos * la cantidad de ingredientes
instance Comida'' Hamburguesa'' where
  comer'' hamburguesa persona = persona {
    peso = peso persona + (3 * genericLength (ingredientes hamburguesa)),
    colesterol = colesterol persona * 1.5
  }
-- las hamburguesas son sabrosas cuando tienen cheddar
  sabrosa'' hamburguesa = elem "cheddar" (ingredientes hamburguesa)
  --sabrosa'' = elem "cheddar".ingredientes

-- la palta aumenta 2 kilos a quien la consume
instance Comida'' Palta'' where
  comer'' _ persona = persona {
    peso = peso persona + 2
  }
--las paltas se repiten comiendolas tres veces  
  repetir'' palta persona = (comer'' palta.comer'' palta.comer'' palta) persona 
--las paltas son sabrosas
  sabrosa'' _ = True

--Como las listas no pueden mezclar diferentes tipos, los almuerzos esta compuestas por comidas del mismo tipo 
almuerzo'' :: [Hamburguesa'']
almuerzo'' = [Hamburguesa'' ["cheddar", "bacon"], Hamburguesa'' ["J","Q","L","T"] ]
--almuerzo'' = [Ensalada'' 1, Ensalada'' 5, Ensalada 2]
--almuerzo'' = [Palta'']

almorzar'' :: Persona -> Persona
almorzar'' persona = foldr comer'' persona almuerzo''

--2) Queremos que todas las comidas se puedan repetir,
-- es decir, que se pueden comer dos veces seguidas, 
-- excepto la palta, que se come tres veces
-- Resuelto en la definicion de la typeClass y cada data

--3) Queremos ver si un almuerzo contiene una comida dada
contieneComida''::(Comida'' a, Eq a) => a->[a]->Bool
contieneComida'' comida comidas = elem comida comidas
--contieneComida = elem

--4) Queremos averiguar si una comida es sabrosa.
-- Resuelto en la definicion de la typeClass y cada data

--5) Queremos averiguar si una comida va a ser disfrutada por alguien.
-- Para quienes pesan más de 100 kg, todas las comidas son disfrutadas, 
-- para los demás, solo son disfrutadas las comidas sabrosas
disfrutar'' :: Comida'' a => a -> Persona -> Bool
disfrutar'' comida alguien = sobrepeso alguien || sabrosa'' comida


-- **************************************************************************
-- COMBINANDO FUNCIONES Y DATA
-- **************************************************************************

-- También es posible combinar diferentes estrategias y definir como funcion
-- alguna de las componentes del data
-- Es util en casos donde hay variedad en los comportamientos, pero la misma
-- estructura de datos 

-- Consideremos el ejemplo anterior, pero ahora  de todas las comidas se conoce 
-- la misma información (kilos y lista de ingredientes). 
-- Los comportamientos son inicialmente los  mismos, pero se prevee 
-- que haya otras alternativas y que un mismo alimento puedan modificarlos

type FormaDeComer = Comida'''->Persona->Persona
type CriterioSabroso = Comida'''->Bool

data Comida''' = Comida''' {
  ingr:: [Ingrediente],
  kil::Kilos,
  formaDeComer:: FormaDeComer,
  criterioSabroso:: CriterioSabroso
}


--1) Queremos que la persona pueda comer distintas comidas. 
-- Existen las ensaladas, las hamburguesas y las paltas, 
-- quiero comer las distintas cosas y cada cosa aporta distinto 
-- al comerla

-- una ensalada de x kilos aporta la mitad de peso para la persona
--     y no agrega colesterol
ensalada'''::FormaDeComer
ensalada''' comida persona = persona {
  peso = peso persona + (kil comida / 2)
}

-- cada hamburguesa tiene una cantidad de ingredientes
-- el colesterol aumenta un 50%
-- el peso aumenta en 3 kilos * la cantidad de ingredientes
hamburguesa'''::FormaDeComer
hamburguesa''' comida persona = persona {
  peso = peso persona + (3 * genericLength (ingr comida)),
  colesterol = colesterol persona * 1.5
}

-- la palta aumenta 2 kilos a quien la consume
palta'''::FormaDeComer
palta''' _ persona  = persona {
  peso = peso persona + 2
}

comer'''::Comida'''->Persona->Persona
comer''' comida persona = (formaDeComer comida) comida persona

almuerzo''' :: [Comida''']
almuerzo''' = [Comida''' [] 1 ensalada''' saborContundente, Comida''' ["cheddar", "bacon"] 0 hamburguesa''' saborChedar, Comida''' [] 0 palta''' siempreSabroso, Comida''' [] 3 ensalada''' siempreSabroso]

almorzar''' :: Persona -> Persona
almorzar''' persona = foldr comer''' persona almuerzo'''

--2) Queremos que todas las comidas se puedan comer dos veces seguidas
repetir''':: Comida''' -> Persona -> Persona
repetir''' comida persona = (comer''' comida.comer''' comida) persona

--3) Queremos ver si un almuerzo contiene una comida dada
instance Eq Comida''' where
  comida1 == comida2 = ingr comida1 == ingr comida2 && kil comida1 == kil comida2

contieneComida'''::Comida'''->[Comida''']->Bool
contieneComida''' comida comidas = elem comida comidas
--contieneComida''' = elem

--4) Queremos averiguar si una comida es sabrosa.
--Se debe agregar otra componente más al data con la correspondiente función
--No necesariamente deben corresponderse con las formas de comer

sabrosa''':: Comida''' -> Bool 
sabrosa''' comida = (criterioSabroso comida) comida

saborChedar, saborContundente, siempreSabroso::CriterioSabroso
saborChedar comida = elem "cheddar" (ingr comida)
saborContundente comida = kil comida > 1
siempreSabroso _ = True

--5) Queremos averiguar si una comida va a ser disfrutada por alguien.
-- Para quienes pesan más de 100 kg, todas las comidas son disfrutadas, 
-- para los demás, solo son disfrutadas las comidas sabrosas
disfrutar''' :: Comida''' -> Persona -> Bool
disfrutar''' comida alguien = sobrepeso alguien || sabrosa''' comida

--Nuevas formas de comer
comidaLigth::FormaDeComer
comidaLigth comida persona 
  | genericLength (ingr comida) >= 2 = persona {peso = peso persona * 1.02}
  | otherwise = persona

comidaInofensiva::FormaDeComer
comidaInofensiva _ p = p

cambiarFormaDeComer::FormaDeComer->Comida'''->Comida'''
cambiarFormaDeComer forma comida = comida {formaDeComer = forma} 
